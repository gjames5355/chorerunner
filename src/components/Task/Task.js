import React, { Component } from 'react';
import ApiService from '../../services/api-service';
import HouseholdContext from '../../contexts/HouseHoldContext';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faPencilAlt,
  faTrashAlt,
  faSave,
} from '@fortawesome/free-solid-svg-icons';

export default class Task extends Component {
  state = {
    editPts: false,
    editTitle: false,
    newPoints: this.props.task.points,
    newTitle: this.props.task.title,
  };

  static contextType = HouseholdContext;

  handleEditTitleClick = () => {
    this.setState({ editTitle: true });
  };

  handleEditPointsClick = () => {
    this.setState({ editPts: true });
  };

  handleTitleUpdate = (id, member_id) => {
    const { household_id } = this.props;
    let reqBody = {
      method: 'title',
      id: id,
      title: this.state.newTitle,
    };

    ApiService.updateTask(household_id, reqBody).then(after => {
      this.setState({ editTitle: false });
      const memberTasks = this.context.tasks;
      memberTasks[member_id].tasks.forEach(task => {
        return parseInt(task.id) === parseInt(id)
          ? (task.title = this.state.newTitle)
          : task.title;
      });
      this.context.setTasks(memberTasks);
    });
  };

  handlePointsUpdate = (id, member_id) => {
    const { household_id } = this.props;
    let reqBody = {
      method: 'points',
      id: id,
      points: this.state.newPoints,
    };

    ApiService.updateTask(household_id, reqBody).then(after => {
      this.setState({ editPts: false });
      const memberTasks = this.context.tasks;
      memberTasks[member_id].tasks.forEach(task => {
        return parseInt(task.id) === parseInt(id)
          ? (task.points = this.state.newPoints)
          : task.points;
      });

      this.context.setTasks(memberTasks);
    });
  };

  handleTitleChange = event => {
    this.setState({ newTitle: event.target.value });
  };

  handlePointsChange = event => {
    this.setState({ newPoints: event.target.value });
  };

  render() {
    const { task, member, handleTaskDelete } = this.props;

    return (
      <li key={task.id} className='assigned-task'>
        <div className='title-points'>
          {this.state.editTitle ? (
            <div className='hhp-title'>
              <input
                className='update-title'
                value={this.state.newTitle}
                onChange={e => {
                  this.handleTitleChange(e);
                }}
              />
              <button
                className='save-title-edit'
                onClick={() =>
                  this.handleTitleUpdate(task.id, member.member_id)
                }
              >
                <FontAwesomeIcon
                  className='save-icon'
                  icon={faSave}
                  size='2x'
                  color=' #b1b1b1'
                />
              </button>
            </div>
          ) : (
            <div className='hhp-title'>
              <span>{task.title}&nbsp;</span>
              <button onClick={() => this.handleEditTitleClick()}>
                <FontAwesomeIcon icon={faPencilAlt} size='lg' color='#b1b1b1' />
              </button>
            </div>
          )}

          {this.state.editPts ? (
            <div className='hhp-points'>
              <input
                className='update-points'
                value={this.state.newPoints}
                onChange={e => {
                  this.handlePointsChange(e);
                }}
              />
              <button
                className='save-points-edit'
                onClick={() =>
                  this.handlePointsUpdate(task.id, member.member_id)
                }
              >
                <FontAwesomeIcon
                  className='save-icon'
                  icon={faSave}
                  size='2x'
                  color=' #b1b1b1'
                />
              </button>
            </div>
          ) : (
            <div className='hhp-points'>
              <span>points: {task.points}</span>
              <button onClick={() => this.handleEditPointsClick()}>
                <FontAwesomeIcon icon={faPencilAlt} size='lg' color='#b1b1b1' />
              </button>
            </div>
          )}
        </div>
        <div className='delete-task-cont'>
          <button
            className='delete-task'
            onClick={() => handleTaskDelete(task.id, member.member_id)}
          >
            <FontAwesomeIcon icon={faTrashAlt} size='2x' color='#b1b1b1' />
            <p>Delete Task</p>
          </button>
        </div>
      </li>
    );
  }
}
