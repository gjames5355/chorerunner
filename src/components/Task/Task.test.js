import React from 'react';
import ReactDOM from 'react-dom';
import Task from './Task.js';
import renderer from 'react-test-renderer';

const task = {
  points: 1,
  title: 'test',
};

const member = {
  member_id: 1,
};

it('renders Task without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Task task={task} member={member} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders Task as expected', () => {
  const tree = renderer
    .create(<Task name='Task' task={task} member={member} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
