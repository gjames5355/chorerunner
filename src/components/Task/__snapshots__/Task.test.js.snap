// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Task as expected 1`] = `
<li
  className="assigned-task"
>
  <div
    className="title-points"
  >
    <div
      className="hhp-title"
    >
      <span>
        test
        Â 
      </span>
      <button
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-pencil-alt fa-w-16 fa-lg "
          color="#b1b1b1"
          data-icon="pencil-alt"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
    <div
      className="hhp-points"
    >
      <span>
        points: 
        1
      </span>
      <button
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-pencil-alt fa-w-16 fa-lg "
          color="#b1b1b1"
          data-icon="pencil-alt"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="delete-task-cont"
  >
    <button
      className="delete-task"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-trash-alt fa-w-14 fa-2x "
        color="#b1b1b1"
        data-icon="trash-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <p>
        Delete Task
      </p>
    </button>
  </div>
</li>
`;
